src/model/reference.ml

```diff
@@ -170,43 +177,95 @@ let tokenize location s : token list =
     let offset = index + 1 in
     let length = started_at - offset in
-    let kind = Some (String.sub s offset length) in
+    let kind = `Prefixed (String.sub s offset length) in
     let location = Location_.in_string s ~offset ~length location in
     let location = Location_.span [ location; identifier_location ] in
     (kind, location)
+  and scan_path started_at index tokens =
+    (* The parsing rules are different for [/]-separated components. [-"".()] are
+       no longer meaningful. *)
+    match s.[index] with
+    | exception Invalid_argument _ -> path_ended started_at index :: tokens
+    | '/' -> scan_path index (index - 1) (path_ended started_at index :: tokens)
+    | _ -> scan_path started_at (index - 1) tokens
+  and path_ended started_at index =
+    let offset = index + 1 in
+    let length = started_at - offset in
+    let identifier = String.sub s offset length in
+    let location = Location_.in_string s ~offset ~length location in
+    { kind = `End_in_slash; identifier; location }
   in
 
   scan_identifier (String.length s) 0 (String.length s - 1) [] |> List.rev
 
-let expected allowed location =
-  let unqualified = "or an unqualified reference" in
+let expected ?(expect_paths = false) allowed location =
+  let unqualified = [ "an unqualified reference" ] in
+  let unqualified =
+    if expect_paths then "a path" :: unqualified else unqualified
+  in
   let allowed =
-    match allowed with
-    | [ one ] -> Printf.sprintf "'%s-' %s" one unqualified
-    | _ ->
-        String.concat ", "
-          (List.map (Printf.sprintf "'%s-'") allowed @ [ unqualified ])
+    String.concat ", " (List.map (Printf.sprintf "'%s-'") allowed @ unqualified)
   in
   expected_err allowed location
 
+(* Parse references that do not contain a [/]. Raises errors and warnings. *)
 let parse whole_reference_location s :
     Paths.Reference.t Error.with_errors_and_warnings =
   let open Paths.Reference in
   let open Names in
+  let rec path components next_token tokens : Hierarchy.t =
+    match (next_token.kind, tokens) with
+    | `End_in_slash, [] when next_token.identifier = "" ->
+        (* {!/identifier} *)
+        (`TAbsolutePath, components)
+    | `End_in_slash, [] when next_token.identifier = "." ->
+        (* {!./identifier} *)
+        (`TRelativePath, components)
+    | `End_in_slash, [ { kind = `End_in_slash; identifier = ""; _ } ]
+      when next_token.identifier = "" ->
+        (* {!//identifier} *)
+        (`TCurrentPackage, components)
+    | `End_in_slash, [] ->
+        (* {!identifier'/identifier} *)
+        (`TRelativePath, next_token.identifier :: components)
+    | `End_in_slash, next_token' :: tokens' ->
+        (* {!path/identifier} *)
+        path (next_token.identifier :: components) next_token' tokens'
+    | (`None | `Prefixed _), _ ->
+        (* This is not really expected *)
+        expected ~expect_paths:true [] next_token.location
+        |> Error.raise_exception
```

Comment:
I think this is impossible, and should be an `assert false`. This also makes `expected` never called with an empty list, and we can put back the `"or "` which makes the error more readable.

Reply:
One way to avoid having this assert false would be to have the `` `End_in_slash `` token contain the whole path, and parse it later.

Reply:
I want to avoid adding structure in the lexer as it's manually written and quite complicated. `expected` with an empty list still gives a valid error message but I agree, let's not put specific error messages in unreachable paths.